DM-48869: Add method to return dataset provenance in a flat dict

timj opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`
- [ ] (if changing dimensions.yaml) make a copy of dimensions.yaml in `configs/old_dimensions`

timj committed:
Add method to return dataset provenance in a flat dict

timj committed:
Write provenance to ArrowAstropy datasets

timj committed:
Add ability to strip and reconstruct provenance from prov dict

timj committed:
Strip provenance from parquet before adding new provenance

erykoff commented on python/lsst/daf/butler/_dataset_provenance.py:
```diff
@@ -82,3 +89,352 @@ def add_extra_provenance(self, dataset_id: uuid.UUID, extra: dict[str, int | flo
+
+        Each input dataset will have the ``id``, ``run``, and ``datasettype``
+        keys as defined above (but no ``dataid`` key) with an ``input N``
+        prefix where ``N`` starts counting at 0.
```
Can you give a fully populated example here?  I'm having trouble wrapping my head around this sentence.

erykoff commented on tests/test_butler.py:
```diff
@@ -2151,10 +2151,31 @@ def test_provenance(self):
+
+        # Check that simple types can be reconstructed with non-standard
+        # separators.
+        prov_dict = prov.to_flat_dict(metric_ref2, prefix="XYZ", sep="😎", simple_types=True)
```
Do we have any other emoji in the codebase?

timj replied:
Yes. Butler uses some for collection names in tests. The developer guide does recommend using unicode in tests.

timj committed:
Add example to DatasetProvenance.to_flat_dict

timj committed:
Improve test coverage

timj commented:
@andy-slac thanks for agreeing to review.

andy-slac commented on python/lsst/daf/butler/_dataset_provenance.py:
```diff
@@ -82,3 +85,94 @@ def add_extra_provenance(self, dataset_id: uuid.UUID, extra: dict[str, int | flo
+                k = k.upper()
+            return f"{start}{k}"
+
+        prov: dict[str, int | float | str | bool | uuid.UUID | DataIdValue] = {}
```
`self.extras` value type is `Any`, does that mean that `prov` can contain `Any` as well if `simple_types` is `False`? (And return type of this method too.)

timj replied:
Yes. I should have changed that to Any when I changed the definition of extras up top (although, it does mean if someone tries to store something in extras that can't be serialized to JSON by Pydantic then things could break). Maybe I should walk back the `extras` scope and use the definition on this line for now (but simple_types is still useful for uuid).

andy-slac commented on python/lsst/daf/butler/_dataset_provenance.py:
```diff
@@ -82,3 +85,94 @@ def add_extra_provenance(self, dataset_id: uuid.UUID, extra: dict[str, int | flo
+        The provenance keys created by this method are defined below. The
+        case of the keys will match the case of the first character of the
+        prefix (defined by whether `str.isupper()` returns true, else they will
+        be lower case).
```
That seems to be a little ad hoc. What if people want upper case keys, but do not want to specify prefix?

timj replied:
Seemed like a highly unlikely scenario but I can add a new parameter that can be used to override the default of using the prefix.

andy-slac commented on python/lsst/daf/butler/_dataset_provenance.py:
```diff
@@ -82,3 +89,372 @@ def add_extra_provenance(self, dataset_id: uuid.UUID, extra: dict[str, int | flo
+            prefix += sep
+        k = sep.join(str(kk) for kk in keys)
+        if use_upper:
+            k = k.upper()
```
This will also uppercase the separator character? I know that it's very likely to be some sort of punctuation, but people are allowed to use any string for separator 🙂

timj replied:
I will make sure that separator does not match `[_\w]`.

andy-slac commented on python/lsst/daf/butler/_dataset_provenance.py:
```diff
@@ -82,3 +89,372 @@ def add_extra_provenance(self, dataset_id: uuid.UUID, extra: dict[str, int | flo
+                flags=re.IGNORECASE,
+            ):
+                _update_matches(match, prefixes, separators)
+            elif match := re.match(r"(.*)\b(id|datasettype|run)$", k, flags=re.IGNORECASE):
```
This assumes that separator should be a non-word character, so underscore cannot be used as separator. Maybe `to_flat_dict` should check that separator is not `\w`?

timj replied:
Will add a check.

andy-slac commented on python/lsst/daf/butler/_dataset_provenance.py:
```diff
@@ -82,3 +89,372 @@ def add_extra_provenance(self, dataset_id: uuid.UUID, extra: dict[str, int | flo
+        return f"{prefix}{k}"
+
+    @staticmethod
+    def _find_prefix_and_sep(prov_dict: Mapping) -> tuple[str, str] | tuple[None, None]:
```
`Mapping[str, Any]` maybe?

andy-slac commented on python/lsst/daf/butler/_dataset_provenance.py:
```diff
@@ -82,3 +89,372 @@ def add_extra_provenance(self, dataset_id: uuid.UUID, extra: dict[str, int | flo
+        return prefixes.pop(), separators.pop()
+
+    @classmethod
+    def _find_provenance_keys_in_flat_dict(cls, prov_dict: Mapping) -> dict[str, str]:
```
Also `Mapping[str, Any]`?

andy-slac commented on python/lsst/daf/butler/_dataset_provenance.py:
```diff
@@ -82,3 +89,372 @@ def add_extra_provenance(self, dataset_id: uuid.UUID, extra: dict[str, int | flo
+            # the input provenance can include extra keys that we cannot
+            # know so have to match solely on INPUT N.
+            found_key = False
+            if re.match(rf"{prefix}input{sep}(\d+){sep}(.*)$", k, flags=re.IGNORECASE):
```
Hmm, what if separator is a backslash?

timj replied:
Thank you.

timj replied:
Backslash now works (now that I escape sep and prefix before using in the regex.

andy-slac commented on python/lsst/daf/butler/_dataset_provenance.py:
```diff
@@ -82,3 +89,372 @@ def add_extra_provenance(self, dataset_id: uuid.UUID, extra: dict[str, int | flo
+                for xk, xv in self.extras[input.id].items():
+                    if simple_types and not isinstance(xv, str | float | int | bool):
+                        xv = str(xv)
+                    prov[_make_key("input", i, xk)] = xv
```
Could `self.extras` contain keys which conflict with other keys, e.g. `datasettype`?

timj replied:
Sigh. Yes. Hmm. This is why I ask you for reviews... I think the fix is to validate add_extra_provenance and disallow run, id, and datasettype as keys.

timj replied:
Validation added.

andy-slac reviewed (approved):
Looks good, a small bunch of comments.

erykoff commented on tests/test_parquet.py:
```diff
@@ -771,12 +773,69 @@ def testBadDataFrameColumnParquet(self):
+            "lsst.butler.datasettype": "data",
+        }
+        self.assertEqual(tab2.meta, expected)
+        null_prov, prov_ref = DatasetProvenance.from_flat_dict(tab2.meta, self.butler)
```
So if anybody were to set a metadata key that was `lsst.butler.something` would this break?  Just curious; it seems fair to have these as reserved keys but I don't know how we could enforce that.

timj replied:
It seems fair to me that only butler can set `lsst.butler.` keys (that's the point of the namespacing). If someone added a `lsst.butler.input 5 id` or something then this would break because the separators would be confused. I think so long as `something` was not `run`, `dataid`, `id`, `datasettype` or `input` then this code wouldn't notice.

erykoff reviewed (approved):
Parquet code looks good.

timj committed:
Escape prefix/sep values as needed and restrict sep values

Can now use "*" as a separator. Previously this broke the
regex for finding the keys. Sep can no longer be alphanumeric
or an underscore.

timj committed:
Restrict usage of provenance reserved keys in extras

timj committed:
Add more explicit type annotation to Mapping parameters

timj committed:
Allow user to control case of keys explicitly

timj committed:
Clarify types that are stored in a provenance dict

timj committed:
Use upper case keys for parquet provenance

In many cases in science pipelines the other keys come
from FITS headers and it looks odd for these to be
lower case.

timj committed:
Fix dtype in test a -> S

DeprecationWarning: Data type alias 'a' was deprecated in NumPy 2.0. Use the 'S' alias instead.

timj committed:
Add news fragment
